File: challenges/2026-08-24-second-deepest-level-sum/solution.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// buildTree builds a tree from LeetCode-style compact level-order values.
// A nil entry in values represents a missing child.
func buildTree(values []*int) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}
	root := &TreeNode{Val: *values[0]}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]
		if i < len(values) {
			if values[i] != nil {
				node.Left = &TreeNode{Val: *values[i]}
				queue = append(queue, node.Left)
			}
			i++
		}
		if i < len(values) {
			if values[i] != nil {
				node.Right = &TreeNode{Val: *values[i]}
				queue = append(queue, node.Right)
			}
			i++
		}
	}
	return root
}

func secondDeepestLevelSum(root *TreeNode) int {
	// TODO: implement
	return 0
}

func iptr(v int) *int { return &v }

func main() {
	type testCase struct {
		values   []*int
		expected int
	}

	tests := []testCase{
		{[]*int{iptr(1), iptr(2), iptr(3), iptr(4), iptr(5)}, 5},
		{[]*int{iptr(10)}, 0},
		{[]*int{iptr(1), iptr(2), nil, iptr(3), nil, iptr(4), nil}, 3},
		{[]*int{}, 0}, // extra: empty tree
		{[]*int{iptr(5), iptr(3), iptr(8)}, 5}, // extra: two levels
	}

	for _, tc := range tests {
		root := buildTree(tc.values)
		got := secondDeepestLevelSum(root)
		status := "FAIL"
		if got == tc.expected {
			status = "PASS"
		}
		fmt.Printf("%s expected=%d got=%d\n", status, tc.expected, got)
	}
}

File: challenges/2026-08-24-second-deepest-level-sum/solution.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val: number, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// buildTree builds a tree from LeetCode-style compact level-order values.
// A null entry represents a missing child.
function buildTree(values: (number | null)[]): TreeNode | null {
  if (values.length === 0 || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length > 0 && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length) {
      const v = values[i++];
      if (v !== null) {
        node.left = new TreeNode(v);
        queue.push(node.left);
      }
    }
    if (i < values.length) {
      const v = values[i++];
      if (v !== null) {
        node.right = new TreeNode(v);
        queue.push(node.right);
      }
    }
  }
  return root;
}

function secondDeepestLevelSum(root: TreeNode | null): number {
  // TODO: implement
  return 0;
}

type Case = [(number | null)[], number];

const tests: Case[] = [
  [[1, 2, 3, 4, 5], 5],
  [[10], 0],
  [[1, 2, null, 3, null, 4, null], 3],
  [[], 0], // extra: empty tree
  [[5, 3, 8], 5], // extra: two levels, second-deepest is the root level
];

for (const [values, expected] of tests) {
  const root = buildTree(values);
  const got = secondDeepestLevelSum(root);
  const status = got === expected ? "PASS" : "FAIL";
  console.log(`${status} root=${JSON.stringify(values)} expected=${expected} got=${got}`);
}
